pkg/eventpersisterops/persister/evaluation_event.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2022 The Bucketeer Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package persister
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/singleflight"
+
+	aoclient "github.com/bucketeer-io/bucketeer/pkg/autoops/client"
+	aodomain "github.com/bucketeer-io/bucketeer/pkg/autoops/domain"
+	"github.com/bucketeer-io/bucketeer/pkg/cache"
+	cachev3 "github.com/bucketeer-io/bucketeer/pkg/cache/v3"
+	featureclient "github.com/bucketeer-io/bucketeer/pkg/feature/client"
+	aoproto "github.com/bucketeer-io/bucketeer/proto/autoops"
+	eventproto "github.com/bucketeer-io/bucketeer/proto/event/client"
+)
+
+const opsEvalKeyPrefix = "autoops:evaluation"
+
+type evalTargetRule struct {
+	id        string
+	clauseIDs []string
+}
+
+type evalEvtUpdater struct {
+	ctx               context.Context
+	featureClient     featureclient.Client
+	autoOpsClient     aoclient.Client
+	eventCounterCache cachev3.EventCounterCache
+	flightgroup       singleflight.Group
+	logger            *zap.Logger
+}
+
+func NewEvalUserCountUpdater(
+	ctx context.Context,
+	featureClient featureclient.Client,
+	autoOpsClient aoclient.Client,
+	eventCounterCache cachev3.EventCounterCache,
+	logger *zap.Logger,
+) Updater {
+	return &evalEvtUpdater{
+		ctx:               ctx,
+		featureClient:     featureClient,
+		autoOpsClient:     autoOpsClient,
+		eventCounterCache: eventCounterCache,
+		logger:            logger,
+	}
+}
+
+func (u *evalEvtUpdater) UpdateUserCounts(ctx context.Context, evt environmentEventMap) map[string]bool {
+	fails := map[string]bool{}
+	for environmentNamespace, events := range evt {
+		for id, event := range events {
+			switch evt := event.(type) {
+			case *eventproto.EvaluationEvent:
+				retriable, err := u.updateUserCount(ctx, environmentNamespace, evt)
+				if err != nil {
+					if err == ErrNoAutoOpsRules || err == ErrAutoOpsRulesNotFound {
+						handledCounter.WithLabelValues(codeNoLink).Inc()
+						u.logger.Debug(
+							"There is no auto ops rules to link the evaluation event",
+							zap.Error(err),
+							zap.String("eventId", id),
+							zap.String("environmentNamespace", environmentNamespace),
+						)
+						continue
```

Comment:
Should we insert `fails[id] = false` at this line?
handledCounter double increments `codeNoLink` and `code.OK`.